Loops: Separate out dynamic_casting concerns from complex overloads.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39263 Kill CPPTypeToScalarType.  It's now subsumed by CPPTypeAndStdComplexToScalarType.
* #39261 Avoid defining bogus CPPTypeAndStdComplexToScalarType<void> by using some decltype tricks.
* #39258 Add dynamic_cast asserts to CPU Loops.
* #39255 Make needs_dynamic_casting multiple-complex-type aware.
* **#39254 Loops: Separate out dynamic_casting concerns from complex overloads.**
* #39246 Avoid a TensorIterator/Loops reinterpret_cast in a test.

dynamic_casting is meant to handle CUDA kernels when the operand dtypes don't match the C++ kernel function types.
This is made more complicated by the current state of complex, which uses thrust::complex, std::complex, c10::complex.

Currently, thrust::complex and std::complex map to need dynamic casting even though we don't actually cast them.
But, making them not need dynamic_cast doesn't work either because certain dynamic_casting optimizations don't work with thrust::complex and (maybe) std::complex.

So, we separate out these concerns so we can iterate on dynamic_casting checks, in particular by applying them to CPU.

This PR should have no functional change.

Differential Revision: [D21788870](https://our.internmc.facebook.com/intern/diff/D21788870)

gchanan committed:
Loops: Separate out dynamic_casting concerns from complex overloads.

dynamic_casting is meant to handle CUDA kernels when the operand dtypes don't match the C++ kernel function types.
This is made more complicated by the current state of complex, which uses thrust::complex, std::complex, c10::complex.

Currently, thrust::complex and std::complex map to need dynamic casting even though we don't actually cast them.
But, making them not need dynamic_cast doesn't work either because certain dynamic_casting optimizations don't work with thrust::complex and (maybe) std::complex.

So, we separate out these concerns so we can iterate on dynamic_casting checks, in particular by applying them to CPU.

This PR should have no functional change.

[ghstack-poisoned]

zasdfgbnm commented on aten/src/ATen/native/cuda/CUDALoops.cuh:
```diff
@@ -64,6 +64,49 @@ static constexpr int launch_bound2 = 4;
+        || std::is_same<thrust::complex<float>, type>::value
+        || std::is_same<thrust::complex<double>, type>::value;
+
+    return c10::guts::if_constexpr<non_c10_complex>([]() {
```
Can we just `return non_c10_complex;`?

gchanan replied:
yes!

zasdfgbnm commented on aten/src/ATen/native/cuda/CUDALoops.cuh:
```diff
@@ -266,7 +312,8 @@ void gpu_kernel_impl(TensorIterator& iter, const func_t& f) {
     }
 
-    if (needs_dynamic_casting<func_t>::check(iter)) {
+    // TODO: can non_c10_complex go through the other path?  Need to verify.
```
There is a `ROCmLoops.cuh` which is a copy-paste of an old version of CUDALoops.cuh, due to compiler bugs in ROCm is blocking the vectorization of GPU loops. It also contains a similar logic. And if are trying to make `needs_dynamic_casting<func_t>::check(iter)` return false for `thrust<-->c10<-->std` casts, you might need to add `|| non_c10_complex` to `ROCmLoops.cuh` as well.

gchanan replied:
oh I see, thanks for pointing this out.

zasdfgbnm commented on c10/util/complex_type.h:
```diff
@@ -99,18 +99,22 @@ namespace c10 {
 //
 // These are implemented by casting to std::complex
 //
+// [NOTE: Complex Operator Unification]
```
Can we put this to the beginning of this file for better visibility?

gchanan replied:
sure, that's a good idea.

gchanan committed:
Update on "Loops: Separate out dynamic_casting concerns from complex overloads."


dynamic_casting is meant to handle CUDA kernels when the operand dtypes don't match the C++ kernel function types.
This is made more complicated by the current state of complex, which uses thrust::complex, std::complex, c10::complex.

Currently, thrust::complex and std::complex map to need dynamic casting even though we don't actually cast them.
But, making them not need dynamic_cast doesn't work either because certain dynamic_casting optimizations don't work with thrust::complex and (maybe) std::complex.

So, we separate out these concerns so we can iterate on dynamic_casting checks, in particular by applying them to CPU.

This PR should have no functional change.

Differential Revision: [D21788870](https://our.internmc.facebook.com/intern/diff/D21788870)

[ghstack-poisoned]

gchanan commented:
@zasdfgbnm: I think I addressed your comments, thanks!

gchanan commented on aten/src/ATen/native/cuda/ROCmLoops.cuh:
```diff
@@ -62,6 +62,49 @@ static constexpr int launch_bound2 = 4;
+        || std::is_same<thrust::complex<float>, type>::value
+        || std::is_same<thrust::complex<double>, type>::value;
+
+    return c10::guts::if_constexpr<non_c10_complex>([]() {
```
whoops, correct.

gchanan replied:
done.

gchanan committed:
Update on "Loops: Separate out dynamic_casting concerns from complex overloads."


dynamic_casting is meant to handle CUDA kernels when the operand dtypes don't match the C++ kernel function types.
This is made more complicated by the current state of complex, which uses thrust::complex, std::complex, c10::complex.

Currently, thrust::complex and std::complex map to need dynamic casting even though we don't actually cast them.
But, making them not need dynamic_cast doesn't work either because certain dynamic_casting optimizations don't work with thrust::complex and (maybe) std::complex.

So, we separate out these concerns so we can iterate on dynamic_casting checks, in particular by applying them to CPU.

This PR should have no functional change.

Differential Revision: [D21788870](https://our.internmc.facebook.com/intern/diff/D21788870)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/CUDALoops.cuh:
```diff
@@ -64,6 +64,45 @@ static constexpr int launch_bound2 = 4;
+  constexpr static bool check() {
+    using traits = function_traits<func_t>;
+    using type = typename traits::template arg<nargs - 1>::type;
+    constexpr bool non_c10_complex =
```
would it make sense to have `is_non_c10_complex_t` in complex_utils.h, similar to how we have `is_complex_t`?

gchanan replied:
if someone needs it we can certainly move it.

My take is we should just have c10_type like we have ztype and map everything that way.
